api/service.go

```diff
@@ -541,148 +585,147 @@ func (s *service) AddFolder(ctx context.Context, req *pb.AddFolderRequest) (*pb.
 		return nil, err
 	}
 
-	pth, err := s.ipfsClient.Unixfs().Add(
-		ctx,
-		ipfsfiles.NewMapDirectory(map[string]ipfsfiles.Node{}),
-		options.Unixfs.Pin(true))
+	bucks, err := s.collections.Buckets.List(ctx, proj.ID)
 	if err != nil {
 		return nil, err
 	}
-
-	folder, err := s.collections.Folders.Create(ctx, pth, req.Name, req.Public, proj.ID)
-	if err != nil {
-		return nil, err
+	list := make([]*pb.GetBucketPathReply, len(bucks))
+	for i, buck := range bucks {
+		list[i], err = s.bucketPathToPb(ctx, buck, path.New(buck.Path), false)
+		if err != nil {
+			return nil, err
+		}
 	}
 
-	return &pb.AddFolderReply{
-		ID:   folder.ID,
-		Path: folder.Path,
-	}, nil
+	return &pb.ListBucketsReply{List: list}, nil
 }
 
-// GetFolder handles a get folder request.
-func (s *service) GetFolder(ctx context.Context, req *pb.GetFolderRequest) (*pb.GetFolderReply, error) {
-	log.Debugf("received get folder request")
-
-	scope, ok := ctx.Value(reqKey("scope")).(string)
-	if !ok {
-		log.Fatal("scope required")
+func (s *service) bucketPathToPb(
+	ctx context.Context,
+	buck *c.Bucket,
+	pth path.Path,
+	followLinks bool,
+) (*pb.GetBucketPathReply, error) {
+	if strings.HasSuffix(pth.String(), ".seed") {
+		return nil, fmt.Errorf("merkledag: not found")
 	}
-	folder, err := s.getFolderWithScope(ctx, req.Name, scope)
+	node, err := s.ipfsClient.Unixfs().Get(ctx, pth)
 	if err != nil {
 		return nil, err
 	}
+	defer node.Close()
 
-	node, err := s.ipfsClient.Unixfs().Get(ctx, path.New(folder.Path))
+	item, err := nodeToBucketItem(pth.String(), node, followLinks)
 	if err != nil {
 		return nil, err
 	}
-	defer node.Close()
 
-	return folderToPbFolder(folder, node)
+	rep := &pb.GetBucketPathReply{
+		Item: item,
+		Root: &pb.BucketRoot{
+			Name:    buck.Name,
+			Path:    buck.Path,
+			Created: buck.Created,
+			Updated: buck.Updated,
+			Public:  buck.Public,
+		},
+	}
+
+	return rep, nil
 }
 
-func folderToPbFolder(folder *c.Folder, node ipfsfiles.Node) (*pb.GetFolderReply, error) {
-	entries, err := dirToPbSlice(node)
+func nodeToBucketItem(pth string, node ipfsfiles.Node, followLinks bool) (*pb.GetBucketPathReply_Item, error) {
+	size, err := node.Size()
 	if err != nil {
 		return nil, err
 	}
-	return &pb.GetFolderReply{
-		ID:        folder.ID,
-		Path:      folder.Path,
-		Name:      folder.Name,
-		Public:    folder.Public,
-		ProjectID: folder.ProjectID,
-		Created:   folder.Created,
-		Entries:   entries,
-	}, nil
-}
-
-func dirToPbSlice(node ipfsfiles.Node) (entries []*pb.GetFolderReply_Entry, err error) {
-	err = ipfsfiles.Walk(node, func(path string, n ipfsfiles.Node) error {
```

Comment:
After playing around with listing paths, I realized a full walk is not really what you want. Just list directories as single items and let the client walk if needed.